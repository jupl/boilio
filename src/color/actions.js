import {createAction} from 'redux-actions'

/**
 * Create action that to move to the next color
 * @type {Function}
 */
export const previousColor = createAction('COLORS_PREVIOUS_COLOR')

/**
 * Create action that to move to the previous color
 * @type {Function}
 */
export const nextColor = createAction('COLORS_NEXT_COLOR')

/**
 * Create action that to move to the next color automatically
 * @type {Function}
 */
export const autoNextColor = createAction('COLORS_AUTO_NEXT_COLOR')
