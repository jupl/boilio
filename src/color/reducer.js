import {handleActions} from 'redux-actions'

// Colors to cycle
const COLORS = ['#39cccc', '#2ecc40', '#ffdc00', '#ff851b']

/** Reducer that handles color related actions */
export default handleActions({
  COLOR_PREVIOUS,
  COLOR_NEXT,
  COLOR_NEXT_AUTO: COLOR_NEXT,
}, COLORS[0])

/**
 * Action handler to go to the previous color. If past the first color, go to
 * the end of the color list.
 * @param {string} state - Current color state
 * @return {string} Previous color state
 */
function COLOR_PREVIOUS(state) {
  const oldIndex = COLORS.indexOf(state)
  const newIndex = (oldIndex + COLORS.length - 1) % COLORS.length
  return COLORS[newIndex]
}

/**
 * Action handler to go to the next color. If past the last color, go to the
 * beginning of the color list.
 * @param {string} state - Current color state
 * @return {string} Next color state
 */
function COLOR_NEXT(state) {
  const oldIndex = COLORS.indexOf(state)
  const newIndex = (oldIndex + 1) % COLORS.length
  return COLORS[newIndex]
}
