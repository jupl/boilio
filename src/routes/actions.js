import {createAction} from 'redux-actions'

/** Create action that to remove current scene */
export const popRoute = createAction('ROUTES_POP')

/** Create action that to switch to a scene */
export const pushRoute = createAction('ROUTES_PUSH', push)

/** Create action to update contents of a scene */
export const refreshRoute = createAction('ROUTES_REFRESH', refresh)

/**
 * Build action data from params when pushRoute action is called
 * @param {string} key - Key identifier for a scene
 * @param {?Object} params - Param data to pass over to scene
 * @return {Object} Action data
 */
function push(key, params) {
  return params != null ? {key, params} : {key}
}

/**
 * Build action data from params when refreshRoute action is called
 * @param {Object} params - Param data to pass over to scene
 * @return {Object} Action data
 */
function refresh(params) {
  return params != null ? {params} : {}
}
