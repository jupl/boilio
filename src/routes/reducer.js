import {handleActions} from 'redux-actions'

/** Reducer that handles color related actions */
export default handleActions({focus}, {})

/**
 * Action handler when scene changes focus
 * @param {Object} state - Current scene information
 * @param {Object} action - Action emitted from router
 * @param {Object} action.scene - Scene/router state
 * @return {Object} New scene information
 */
function focus(state, {scene}) {
  return JSON.parse(JSON.stringify(scene))
}
