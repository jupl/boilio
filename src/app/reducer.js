import {combineReducers} from 'redux'
import color from '../color/reducer'

/** Combination of reducer subsets as a single reducer for Redux */
export default combineReducers({color})
