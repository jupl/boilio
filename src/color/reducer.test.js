import reducer from './reducer'
import * as actions from './actions'

describe('Color reducer', () => {
  // Inital state
  const state0 = reducer(undefined, {type: ''})

  it('should start with initial state', () => {
    expect(state0).toMatchSnapshot()
  })

  it('should handle previous color', () => {
    const action = actions.previousColor()
    const state1 = reducer(state0, action)
    const state2 = reducer(state1, action)
    expect(state1).toMatchSnapshot()
    expect(state2).toMatchSnapshot()
  })

  it('should handle next color', () => {
    const action = actions.nextColor()
    const state1 = reducer(state0, action)
    const state2 = reducer(state1, action)
    expect(state1).toMatchSnapshot()
    expect(state2).toMatchSnapshot()
  })

  it('should handle auto next color', () => {
    const action = actions.autoNextColor()
    const state1 = reducer(state0, action)
    const state2 = reducer(state1, action)
    expect(state1).toMatchSnapshot()
    expect(state2).toMatchSnapshot()
  })
})
