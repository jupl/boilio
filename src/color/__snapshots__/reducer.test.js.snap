exports[`Color reducer should handle auto next color 1`] = `"#2ecc40"`;

exports[`Color reducer should handle auto next color 2`] = `"#ffdc00"`;

exports[`Color reducer should handle next color 1`] = `"#2ecc40"`;

exports[`Color reducer should handle next color 2`] = `"#ffdc00"`;

exports[`Color reducer should handle previous color 1`] = `"#ff851b"`;

exports[`Color reducer should handle previous color 2`] = `"#ffdc00"`;

exports[`Color reducer should start with initial state 1`] = `"#39cccc"`;
